### PR TITLE
Loki telemetry adapter

### PR DESCRIPTION
**Pull request by patuwwy**
**What?**  

- Add `@scramjet/telemetry` package
  -  with Loki Telemetry Adapter
- Add `--no-telemetry` to switch telemetry off
- Report host start (w/ host size)
- Report seq upload (w/ language, w/ host size)
- Report instance start (w/ host size, id)
- Repoer instance end/error (w/ host size, execution time)
- Show telemetry message with status after installing `@scramjet/sth`
- Show telemetry message when sth is started and telemetry is active

**Needs to provide desired Loki url to be merged!**

**Review checks:**

These aspects need to be checked by the reviewer:

- [x] Verify and confirm operation (please post a screenshot) 
- [x] All STH tests pass
- [ ] All [Scramjet Cloud Platform](https://docs.scramjet.org/platform) tests pass
- [ ] Documentation is updated or no changes

### CODE REVIEW

**Comment by BKscramjet** [comment]
Task linked: [SHN-668 Loki Adapter](https://app.clickup.com/t/24308805/SHN-668)

File: packages/telemetry/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { TelemetryAdaptersConfig } from "@scramjet/types";
+import { LokiAdapter } from "./adapters/loki";
```
Comment (MichalCz) [inline packages/telemetry/src/index.ts:2]: Food for thoughts here: we can't do tree-shaking when we don't use the specific adapter - this is equally true for for instance kubernetes adapter - it's not loaded, so memory is not taken there, but we still need to ship it with the libs... Takes up disk space at least, the above takes also memory.

**Comment by BKscramjet** [comment]
Task linked: [SHN-603 Add telemetry reporting module](https://app.clickup.com/t/24308805/SHN-603)

**Comment by BKscramjet** [comment]
Task linked: [SHN-604 Add dummy telemetry adapter](https://app.clickup.com/t/24308805/SHN-604)

File: packages/telemetry/src/index.ts (hunk 1 above)
Comment (patuwwy) [reply packages/telemetry/src/index.ts]: we can split these modules, `adapters` as well, into smaller ones and install them at runtime (at startup) if they are needed

**Comment by patuwwy** [comment]
```
Common labels: host | Line limit: 1000 (35 returned) | Total bytes processed: 754 B
2022-08-23 01:44:58 Instance ended 
  Log labels
    code	        0
    executionTime	22.266
    hostSize	        xl
    id	                dd359996-81b6-4251-9500-c5c7aa77ab07
    level	        info
    module	        host
  Detected fields
    Time	        1661211898275
    labels	        [object Object]
    tsNs	        1661211898275000000

2022-08-23 01:44:36 Instance started 
  Log labels
    hostSize	        xl
    id	                dd359996-81b6-4251-9500-c5c7aa77ab07
    language	        js
    level	        info
    module	        host
  Detected fields
    Time	        1661211876013
    labels	        [object Object]
    tsNs	        1661211876013000000

2022-08-23 01:44:18 Sequence uploaded 
  Log labels
    hostSize	        xl
    language	        js
    level	        info
    module	        host
  Detected fields
    Time	        1661211858694
    labels	        [object Object]
    tsNs	        1661211858694000000

2022-08-23 01:43:00 Host started 
  Log labels
    hostSize	        xl
    level	        info
    module	        host
  Detected fields
    Time	        1661211780557
    labels	        [object Object]
    tsNs	        1661211780557000000
```

**Comment by BKscramjet** [comment]
Task linked: [SHN-606 Add telemetry calls to host](https://app.clickup.com/t/24308805/SHN-606)

File: packages/adapters/src/process-sequence-adapter.ts [hunk 2]
```diff
@@ -42,6 +44,7 @@ async function getRunnerConfigForStoredSequence(sequencesRoot: string, id: strin
         author: validPackageJson.author || "",
         keywords: validPackageJson.keywords || [],
         repository: validPackageJson.repository || "",
+        language: language || engines?.node ? "js" : "unknown"
```
Comment (MichalCz) [inline packages/adapters/src/process-sequence-adapter.ts:47]: Patryk, why is this different to kube/docker adapters?

File: packages/host/src/lib/host.ts [hunk 3]
```diff
@@ -1011,5 +1041,38 @@ export class Host implements IComponent {
...
+                }).question("Do You consent to the transmission of anonymous usage data to help us improve Transform Hub? (y/n) ", (reply: string) => {
+                    this.logger.info("User replied", reply);
+
+                    this.config.telemetry.status = ["yes", "y"].includes(reply.trim()) ? "on" : "off";
+                    resolve();
+                });
+            });
+        }
+
+        if (this.config.telemetry.status === "on") {
```
Comment (MichalCz) [inline packages/host/src/lib/host.ts:1060]: Patryk, we were supposed to add a message about telemetry and on how to switch it off.

**Review by MichalCz — APPROVED** [review]

File: packages/sth/package.json [hunk 4]
```diff
@@ -14,7 +14,8 @@
     "build:docker": "docker build -t scramjetorg/sth:$(git rev-parse HEAD) -f Dockerfile ../../",
     "clean": "rm -rf ./dist .bic_cache",
     "test": "echo no tests yet -- # npm run test:ava",
-    "test:ava": "ava"
+    "test:ava": "ava",
+    "postinstall": "./src/postinstall.ts"
```
Comment (MichalCz) [inline packages/sth/package.json:18]: Hmmm... Will this change it's location when we build the package? I'd use clean `js` instead.

**Review by alicja-gruzdz — COMMENTED** [review]
Test 1/2: running Instances on STH provided by @patuwwy registered the events correctly.
Screenshot from 2022-08-23 11-02-48

Test 2/2 in progress...

**Comment by alicja-gruzdz** [comment]
It will a very useful feature. 

One suggestion is to add an informative flag that would help to differentiate our internal uses of STH from the uses of external users.

**Review by alicja-gruzdz — APPROVED** [review]
Test 2/2: Running local STH:
Screenshot from 2022-08-23 13-00-45

Entry saved in Grafana:
Screenshot from 2022-08-23 13-01-05

Telemetry flags off:

Screenshot from 2022-08-23 13-04-16

Screenshot from 2022-08-23 13-05-44

Starting of STH was not registered, still there is only one entry on the list (the previously started STH):

Screenshot from 2022-08-23 13-06-03

**Comment by alicja-gruzdz** [comment]
We need to test again after the release / cc: @a-tylenda

File: packages/adapters/src/process-sequence-adapter.ts (hunk 2 above)
Comment (patuwwy) [reply packages/adapters/src/process-sequence-adapter.ts:47]: i forgot to align changes across sequence adapters. fixed

File: packages/host/src/lib/host.ts (hunk 3 above)
Comment (alicja-gruzdz) [reply packages/host/src/lib/host.ts]: Either there's one or Patryk added but in any case it displays nicely :) So I mark it as resolved. 
Screenshot from 2022-09-08 11-25-38

File: packages/sth/package.json (hunk 4 above)
Comment (alicja-gruzdz) [reply packages/sth/package.json]: thanks, it has been modified by @patuwwy to:
```
  "postBuildOverride": {
    "scripts": {
      "postinstall": "./postinstall.js"
    }
```
so I mark it as resolved.

**Comment by alicja-gruzdz** [comment]
verified by @a-tylenda, thanks!

**Pull request merged**